components/camel-zipfile/src/main/java/org/apache/camel/processor/aggregate/zipfile/ZipAggregationStrategy.java

```diff
@@ -178,81 +191,59 @@ public Exchange aggregate(Exchange oldExchange, Exchange newExchange) {
                 // do not try to append empty data
                 if (buffer.length > 0) {
                     String entryName = useFilenameHeader ? newExchange.getIn().getHeader(Exchange.FILE_NAME, String.class) : newExchange.getIn().getMessageId();
-                    addEntryToZip(zipFile, entryName, buffer, buffer.length);
-                    GenericFile<File> genericFile = FileConsumer.asGenericFile(
-                            zipFile.getParent(), zipFile, Charset.defaultCharset().toString(), false);
-                    genericFile.bindToExchange(answer);
+                    addEntryToZip(zipFile, entryName, buffer, charset);
                 }
             } catch (Exception e) {
                 throw new GenericFileOperationFailedException(e.getMessage(), e);
             }
         }
-        
+
+        GenericFile<File> genericFile = FileConsumer.asGenericFile(zipFile.getParent(), zipFile, charset, false);
+        genericFile.bindToExchange(answer);
+
         return answer;
     }
-    
-    private static void addFileToZip(File source, File file, String fileName) throws IOException {
-        File tmpZip = Files.createTempFile(source.getName(), null).toFile();
-        tmpZip.delete();
-        if (!source.renameTo(tmpZip)) {
-            throw new IOException("Could not make temp file (" + source.getName() + ")");
-        }
-        byte[] buffer = new byte[8192];
 
-        FileInputStream fis = new FileInputStream(tmpZip);
-        ZipInputStream zin = new ZipInputStream(fis);
-        ZipOutputStream out = new ZipOutputStream(new FileOutputStream(source));
-
-        try {
-            InputStream in = new FileInputStream(file);
-            out.putNextEntry(new ZipEntry(fileName == null ? file.getName() : fileName));
-            for (int read = in.read(buffer); read > -1; read = in.read(buffer)) {
-                out.write(buffer, 0, read);
+    private static void newZipFile(File zipFile) throws URISyntaxException, IOException {
+        if (zipFile.exists()) {
```

Comment:
These 2 if can maybe be combined into 1 if with `&&`

Reply:
True, I will do that. I have mixed artifacts in my local maven. I need to rebuild project and then will commit this improvement (approx 1 hour)

Reply:
Combined in commit 9afa571